proposals/0327-actor-initializers.md

```diff
@@ -42,126 +49,129 @@ The protection is achieved by _isolating_ the mutable state of each actor instan
-      fatalError("state changed by another thread!")
+  func click(_ times: n) {
+    for _ in 0<..n {
+      self.count += 1 
     }
   }
 
-  func tick() {
-    self.counter = self.counter + 1
+  deinit {
+    let current = count
+    
+    Task { await self.click(1000000) } // This might keep `self` alive after the `deinit` too!
+
+    assert(current == count)    // 💥 Might fail due to data-race!
   }
 }
 ```
 
-This example exhibits a race because `self`, once fully-initialized, is ready to provide isolated access to its members, i.e., it does *not* start in a reserved state. Isolated access is obtained by "hopping" to the executor corresponding to `self` from an asynchronous function. But, because `init` is synchronous, a hop to `self` fundamentally cannot be performed. Thus, once `self` is initialized, the remainder of the `init` is subject to the kind of data race that actors are meant to eliminate.
-
-If the `init` in the previous example were only changed to be `async`, this data race still does not go away. The existing implementation does not perform a hop to `self` in such initializers, even though it now could to prevent races. This is not just a bug that has a straightforward fix, because if an asynchronous actor `init` were isolated to the `@MainActor`: 
+There is another more subtle scenario in the `deinit` of global-actor isolated classes (GAICs). A GAIC is similar to an actor that shares its persistent executor with other instances. When an actor or GAIC's executor is not exclusively owned by the type's instance, then it is not safe to access a non-Sendable stored property from the `deinit`:
 
 ```swift
-class ConnectionStatusDelegate {
-  @MainActor
-  func connectionStarting() { /**/ }
-
-  @MainActor
-  func connectionEstablished() { /**/ }
+class NonSendableAhmed { 
+  var state: Int = 0
 }
 
-actor ConnectionManager {
-  var status: ConnectionStatusDelegate
-  var connectionCount: Int
-
-  @MainActor
-  init(_ sts: ConnectionStatusDelegate) async {
-    // --- on MainActor --
-    self.status = sts
-    self.status.connectionStarting()
-    self.connectionCount = 0
-    // --- self fully-initialized here ---
-    
-    // ... connect ...
-    self.status.connectionEstablished()
+@MainActor
+class Maria {
+  let friend: NonSendableAhmed
+
+  init() {
+    self.friend = NonSendableAhmed()
+  }
+
+  init(sharingFriendOf otherMaria: Maria) {
+    // While the friend is non-Sendable, this initializer and
+    // and the otherMaria are isolated to the MainActor. That is,
+    // they share the same executor. So, it's OK for the non-Sendable value
+    // to cross between otherMaria and self.
+    self.friend = otherMaria.friend
+  }
+
+  deinit {
+    friend.state += 1   // 💥 the deinit is not isolated to the MainActor,
+                        // so this mutation can happen concurrently with other
+                        // accesses to the same underlying instance of 
+                        // NonSendableAhmed.
```

Comment:
This example confuses me a bit since its not really about the deinit -- it's just about any mutation made to friend _at all_ since it's not sendable it never is truly safe 🤔

Reply:
But I get the point this is trying to make in general hmmm

Reply:
The shared `friend` can be safely mutated from an isolated method, but not from the `deinit` because it does not have isolation.